File: depth_first_search/1457_pseudo_palindromic_paths_in_a_binary_tree/1457_pseudo_palindromic_paths_in_a_binary_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

        
class Solution:
    def pseudoPalindromicPaths (self, root: TreeNode) -> int:
        s = set()

        def dfs(root):

            if not root: return 0

            if root.val in s:
                s.discard(root.val)
            else:
                s.add(root.val)

            res = dfs(root.left) + dfs(root.right)

            if not root.left and not root.right:
                res += len(s)<=1

            if root.val in s:
                s.discard(root.val)
            else:
                s.add(root.val)

            return res

        return dfs(root)

File: depth_first_search/1457_pseudo_palindromic_paths_in_a_binary_tree/1457_pseudo_palindromic_paths_in_a_binary_tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pseudoPalindromicPaths(root *TreeNode) int {
	if root == nil {
		return 0
	}
	var ans int
	var h func(node *TreeNode, temp int)
	h = func(node *TreeNode, temp int) {
		temp ^= (1 << node.Val)
		if node.Left == nil && node.Right == nil {
			if temp == 0 || temp&(temp-1) == 0 {
				ans++
			}
		}

		if node.Left != nil {
			h(node.Left, temp)
		}

		if node.Right != nil {
			h(node.Right, temp)
		}

	}

	h(root, 0)
	return ans
}
